File: gripss/src/main/java/com/hartwig/hmftools/gripss/links/AlternatePath.java
package com.hartwig.hmftools.gripss.links;

import static com.hartwig.hmftools.gripss.links.Link.LINK_TYPE_PAIR;
import static com.hartwig.hmftools.gripss.links.TransitiveLink.TRANS_LINK_PREFIX;

import java.util.List;
import java.util.StringJoiner;
import java.util.stream.Collectors;

import com.hartwig.hmftools.gripss.common.Breakend;

public class AlternatePath
{
    // an alternative path is modelled as the pair of breakends from the same SV, in a specific order
    // and the set of links made by other SVs that start and end at the same locations
    public final Breakend First;
    public final Breakend Second;
    public final List<Link> Links;

    public AlternatePath(final Breakend first, final Breakend second, final List<Link> links)
    {
        First = first;
        Second = second;
        Links = links;
    }

    public List<Link> transitiveLinks()
    {
        return Links.stream().filter(x -> x.Id.startsWith(TRANS_LINK_PREFIX)).collect(Collectors.toList());
    }

    public String pathString()
    {
        StringJoiner sj = new StringJoiner("");

        for(int i = 0; i < Links.size(); ++i)
        {
            Link link = Links.get(i);

            if (i == 0)
                sj.add(link.breakendStart().VcfId);

            if (link.Id.equals(LINK_TYPE_PAIR))
                sj.add("-");
            else
                sj.add(String.format("<%s>", link.Id));

            sj.add(link.breakendEnd().VcfId);
        }

        return sj.toString();
    }

    public String toString()
    {
        return String.format("breaks(%s - %s) links(%d)", First, Second, Links.size());
    }

}

File: gripss_kt/src/main/kotlin/com/hartwig/hmftools/gripss/link/AlternatePath.kt
package com.hartwig.hmftools.gripss.link

import com.hartwig.hmftools.gripss.store.LinkStore
import com.hartwig.hmftools.gripss.store.VariantStore
import org.apache.logging.log4j.LogManager
import java.util.*
import kotlin.collections.HashSet

data class AlternatePath(val vcfId: String, val mateId: String, val path: List<Link>) {

    companion object {
        private val logger = LogManager.getLogger(this::class.java)

        operator fun invoke(assemblyLinkStore: LinkStore, variantStore: VariantStore): Collection<AlternatePath> {

            val failed = HashSet<String>()
            val result = HashMap<String, AlternatePath>()
            val transitiveLink = TransitiveLink(assemblyLinkStore, variantStore)

            for (variant in variantStore.selectAll()) {
                if (variant.mateId != null && !result.keys.contains(variant.mateId) && !failed.contains(variant.mateId)) {
                    val links = transitiveLink.transitiveLink(variant)
                    if (links.isNotEmpty()) {
                        val alternatePath = AlternatePath(variant.vcfId, variant.mateId, links)
                        val reverseAlternatePath = AlternatePath(variant.mateId, variant.vcfId, links.map { x -> x.reverse() }.reversed())
                        result[variant.vcfId] = alternatePath
                        result[variant.mateId] = reverseAlternatePath
                        logger.debug("Found alternate mapping of $variant CIPOS:${variant.confidenceInterval} IMPRECISE:${variant.imprecise} -> ${alternatePath.pathString()}")
                    } else {
                        failed.add(variant.vcfId)
                    }
                }
            }

            return result.values
        }
    }

    fun size(): Int = path.size

    fun pathVcfIds(): List<String> {
        return path.map { x -> x.vcfId } + path[path.size - 1].otherVcfId
    }

    fun transitiveLinks(): List<Link> {
        return path.filter { x -> x.link.startsWith("trs") }
    }

    fun pathString(): String {
        val stringJoiner = StringJoiner("")

        for (i in path.indices) {
            if (i == 0) {
                stringJoiner.add(path[i].vcfId)
            }

            val link = path[i]
            if (link.link == "PAIR") {
                stringJoiner.add("-")
            } else {
                stringJoiner.add("<${link.link}>")
            }

            stringJoiner.add(link.otherVcfId)
        }

        return stringJoiner.toString()
    }
}